Marlin/src/feature/bedlevel/bdl/bdl.cpp

```diff
@@ -42,151 +43,200 @@
         tmp = BD_I2C_SENSOR.BD_i2c_read();
-        SERIAL_ECHOLNPGM("Calibrate data:", i, ",", tmp & 0x3FF, ", check:", BD_I2C_SENSOR.BD_Check_OddEven(tmp));
-        safe_delay(500);
+        SERIAL_ECHOLNPGM("Calibrate data (", i, ") = ", tmp & 0x3FF);
+        (void)check(tmp, i == 0);
+        safe_delay(50);
       }
       config_state = 0;
       BD_I2C_SENSOR.BD_i2c_write(CMD_END_READ_CALIBRATE_DATA);
-      safe_delay(500);
+      safe_delay(50);
     }
     else if (config_state <= -6) {   // Start Calibrate
-      safe_delay(100);
+      safe_delay(10);
       if (config_state == -6) {
-        //BD_I2C_SENSOR.BD_i2c_write(1019); // begin calibrate
-        //delay(1000);
-        gcode.stepper_inactive_time = SEC_TO_MS(60 * 5);
-        gcode.process_subcommands_now(F("M17 Z"));
-        gcode.process_subcommands_now(F("G1 Z0.0"));
-        z_pose = 0;
+        REMEMBER(gsit, gcode.stepper_inactive_time, SEC_TO_MS(60 * 5));
+        SERIAL_ECHOLNPGM("c_z0:", planner.get_axis_position_mm(Z_AXIS), "-", pos_zero_offset);
+
+        // Move the z axis instead of enabling the Z axis with M17
+        // TODO: Use do_blocking_move_to_z for synchronized move.
+        current_position.z = 0;
+        sync_plan_position();
+        gcode.process_subcommands_now(F("G1Z0.05"));
         safe_delay(1000);
+        gcode.process_subcommands_now(F("G1Z0.00"));
+        safe_delay(1000);
+        current_position.z = 0;
+        sync_plan_position();
+        //safe_delay(1000);
+
+        while ((planner.get_axis_position_mm(Z_AXIS) - pos_zero_offset) > 0.00001f) {
+          safe_delay(200);
+          SERIAL_ECHOLNPGM("c_z6:", planner.get_axis_position_mm(Z_AXIS));
+        }
+        zpos = 0.00001f;
+        safe_delay(100);
         BD_I2C_SENSOR.BD_i2c_write(CMD_START_CALIBRATE); // Begin calibrate
-        SERIAL_ECHOLNPGM("Begin calibrate");
-        safe_delay(2000);
+        SERIAL_ECHOLNPGM("BD Sensor Calibrating...");
+        safe_delay(200);
         config_state = -7;
       }
-      else if (planner.get_axis_position_mm(Z_AXIS) < 10.0f) {
-        if (z_pose >= MAX_BD_HEIGHT) {
+      else if ((planner.get_axis_position_mm(Z_AXIS) - pos_zero_offset) < 10.0f) {
+        if (zpos >= MAX_BD_HEIGHT) {
           BD_I2C_SENSOR.BD_i2c_write(CMD_END_CALIBRATE); // End calibrate
-          SERIAL_ECHOLNPGM("End calibrate data");
-          z_pose = 7;
+          SERIAL_ECHOLNPGM("BD Sensor calibrated.");
+          zpos = 7.0f;
           config_state = 0;
           safe_delay(1000);
         }
         else {
-          float tmp_k = 0;
-          char tmp_1[30];
-          sprintf_P(tmp_1, PSTR("G1 Z%d.%d"), int(z_pose), int(int(z_pose * 10) % 10));
+          char tmp_1[32];
+          // TODO: Use prepare_internal_move_to_destination to guarantee machine space
+          sprintf_P(tmp_1, PSTR("G1Z%d.%d"), int(zpos), int(zpos * 10) % 10);
           gcode.process_subcommands_now(tmp_1);
```

Comment:
Although running G-code can work, it is better to use the internal motion API to guarantee machine space and absolute movement.

Reply:
agree, and fortunately this calibration code are only run in the first time after user mounted the sensor.